### PR TITLE
# 76 environment and agent

### PR DESCRIPTION
**Pull request by AlexKay28**
- add environment
- add agend
- implement RL as separate microservice

### CODE REVIEW

**Review by Alkhimovmv — APPROVED** [review]

File: backend/utils/db_controller.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+                            "phrase_from": idx_to + 1,
+                            "phrase_to": idx_from + 1,
+                            "average_success": 1.0,
+                        },
+                    ],
+                    ignore_index=True,
+                )
+
+        # upload data to base
+        dataframe.to_sql(
```
Comment (alantishin) [inline backend/utils/db_controller.py:99]: its better to run this three queries inside transaction
especially with if_exists="replace"
if something go wrong consistency will be broken

File: backend/api/update_transitions.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    Update transition success matrix using actions table
+    """
+    try:
+        # auth token
+        session_token = request.headers.get("session_token")
+
+        # work with provided file
+        db_controller.update_transitions_global()
+        message = "transitions updated!"
+        return jsonify(
```
Comment (alantishin) [inline backend/api/update_transitions.py:22]: All this return jsonify
seems like something that can be moved to separate functions and reused repeatedly 

like this 
`def response_message(message, code=200):
   return jsonify(
            {
                "status": 200,
                "message": message,
            }
        )`

here and in other similar 
code becomes in one line
`return response_message(message)` 

returning errors are same

File: backend/utils/db_controller.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        transition_success_table.to_sql(
+            "transition_success",
+            self.dbconnector.engine,
+            schema="public",
+            if_exists="replace",
+            index=False,
+            method="multi",
+        )
+
+        return 0
```
Comment (alantishin) [inline backend/utils/db_controller.py:124]: it looks strange

File: backend/utils/session_object.py [hunk 4]
```diff
@@ -6,14 +6,18 @@
...
+
 from dbase import db
 from dbase.users import UserAuthorized, UserAnon
 from dbase.actions import Action
 from dbase.phrases import Phrase
-from dbase.database_connector import DatabaseConnector
 
 N_MAX_USERS = 25
 
+RL_SERVICE_URL = "http://localhost:6768/get_pair"
```
Comment (alantishin) [inline backend/utils/session_object.py:19]: hardcode is bad )

Comment (AlexKay28) [reply backend/utils/session_object.py:19]: i have added this as environment variable

File: backend/utils/db_controller.py (hunk 1 above)
Comment (AlexKay28) [reply backend/utils/db_controller.py:109]: this is kind of hard reset of the main tables. it doesn't update anything already existed

File: backend/api/update_transitions.py (hunk 2 above)
Comment (AlexKay28) [reply backend/api/update_transitions.py:22]: well, nice recommentation! 
it going to be a separate issue

**Review by alantishin — APPROVED** [review]

**Pull request merged**